src/daemons/linuxd/src/linuxd/mod.rs

```diff
@@ -677,73 +667,56 @@ impl<'a> LinuxDaemon<'a> {
                     return message;
                 }
 
-                let mut length_buffer: [u8; mem::size_of::<u32>()] = [0u8; mem::size_of::<u32>()];
-                match conn.read_exact(&mut length_buffer) {
-                    Ok(_) => {
-                        let length: u32 = u32::from_le_bytes(length_buffer);
-                        if length == 0 {
-                            debug!("read 0 bytes from the gateway");
-                            ReadResponse::build(source, 0, [0u8; ReadResponse::BUFFER_SIZE])
+                match conn.read_message_from_gateway() {
+                    Ok(message) => {
+                        let count: usize = message.len();
+                        debug!("read {count} bytes from the gateway");
+
+                        // Truncate read request to fit in the response buffer.
+                        let read_count: usize = if count > request.count as usize {
+                            warn!(
+                                "handle_read_request(): truncating payload (requested={}, \
+                                 actual={count})",
+                                { request.count },
+                            );
+                            request.count as usize
```

Comment:
Instead of logging a warning message we should fail. If failing is not possible, we should panic. People don't fix warning messages, but people do fix critical errors.

Reply:
This is not an error. This warning only means that we will chunk the input message into different messages to the micro VM.